test/satellite/process.tags.test.ts

```diff
@@ -0,0 +1,790 @@
+  app: 'test',
+  env: 'default',
+}
+
+test.beforeEach(async (t) => {
+  await mkdir('.tmp', { recursive: true })
+  const dbName = `.tmp/test-${randomValue()}.db`
+  const db = new Database(dbName)
+  const adapter = new DatabaseAdapter(db)
+  const migrator = new BundleMigrator(adapter, migrations)
+  const notifier = new MockNotifier(dbName)
+  const client = new MockSatelliteClient()
+  const console = new MockConsoleClient()
+  const satellite = new SatelliteProcess(
+    dbName,
+    adapter,
+    migrator,
+    notifier,
+    client,
+    console,
+    satelliteConfig,
+    opts
+  )
+
+  const tableInfo = initTableInfo()
+  const timestamp = new Date().getTime()
+
+  const runMigrations = async () => {
+    await migrator.up()
+  }
+
+  t.context = {
+    dbName,
+    db,
+    adapter,
+    migrator,
+    notifier,
+    client,
+    runMigrations,
+    satellite,
+    tableInfo,
+    timestamp,
+  }
+})
+
+test.afterEach.always(async (t) => {
+  const { dbName, satellite } = t.context as any
+
+  await removeFile(dbName, { force: true })
+  await removeFile(`${dbName}-journal`, { force: true })
+
+  await satellite.stop()
+})
+
+test('basic rules for setting tags', async (t) => {
+  const { adapter, runMigrations, satellite } = t.context as any
+  await runMigrations()
+
+  await satellite._setAuthState()
+  const clientId = satellite['_authState']['clientId']
+
+  await adapter.run({
+    sql: `INSERT INTO parent(id, value, other) VALUES (1, 'local', null)`,
+  })
+
+  const txDate1 = await satellite._performSnapshot()
+  let shadow = await satellite._getOplogShadowEntry()
+  t.is(shadow.length, 1)
+  t.is(shadow[0].tags, genEncodedTags(clientId, [txDate1]))
+
+  await adapter.run({
+    sql: `UPDATE parent SET value = 'local1', other = 'other1' WHERE id = 1`,
+  })
+
+  const txDate2 = await satellite._performSnapshot()
+  shadow = await satellite._getOplogShadowEntry()
+  t.is(shadow.length, 1)
+  t.is(shadow[0].tags, genEncodedTags(clientId, [txDate2]))
+
+  await adapter.run({
+    sql: `UPDATE parent SET value = 'local2', other = 'other2' WHERE id = 1`,
+  })
+
+  const txDate3 = await satellite._performSnapshot()
+  shadow = await satellite._getOplogShadowEntry()
+  t.is(shadow.length, 1)
+  t.is(shadow[0].tags, genEncodedTags(clientId, [txDate3]))
+
+  await adapter.run({
+    sql: `DELETE FROM parent WHERE id = 1`,
+  })
+
+  const txDate4 = await satellite._performSnapshot()
+  shadow = await satellite._getOplogShadowEntry()
+  t.is(shadow.length, 0)
+
+  let entries = await satellite._getEntries()
+  //console.log(entries)
+  t.is(entries[0].clearTags, encodeTags([]))
+  t.is(entries[1].clearTags, genEncodedTags(clientId, [txDate1]))
+  t.is(entries[2].clearTags, genEncodedTags(clientId, [txDate2]))
+  t.is(entries[3].clearTags, genEncodedTags(clientId, [txDate3]))
+
+  t.not(txDate1, txDate2)
+  t.not(txDate2, txDate3)
+  t.not(txDate3, txDate4)
+})
+
+// In this test we take three snapshots for three separate transactions.
```

Comment:
Please remove the long explanation (which was meant to you) and adapt the test name accordingly with other tests you created.

Reply:
removed